src/compiler/expressionToTypeNode.ts

```diff
@@ -814,7 +814,7 @@ export function createSyntacticTypeNodeBuilder(
     }
     function typeFromProperty(node: PropertyDeclaration | PropertySignature | JSDocPropertyTag, symbol: Symbol, context: SyntacticTypeNodeBuilderContext) {
         const declaredType = getEffectiveTypeAnnotationNode(node);
-        const requiresAddingUndefined = resolver.requiresAddingImplicitUndefined(node, symbol, context.enclosingDeclaration);
+        const requiresAddingUndefined = strictNullChecks && resolver.requiresAddingImplicitUndefined(node, symbol, context.enclosingDeclaration);

```

Comment:
I have a bad feeling that making isolated declaration emit _more_ dependent on compiler options is a bad idea... Theoretically adding undefined here isn't a big deal because it "doesn't exist" when handled by the checker.

How does this end up fixing the assert?

Reply:
Hm, this isn't that relevant to ID (from what I understand). In the example, we want to generate types for an output of a factory function and ID requires explicit annotations - so it wouldn't even run into this problem. 

The issue arises because you have a homomorphic mapped type (that preserves modifiers) running over `prop?: undefined`. So it ran into this code in `requiresAddingImplicitUndefined`:
```ts
                switch (declaration.kind) {
                    case SyntaxKind.PropertyDeclaration:
                    case SyntaxKind.PropertySignature:
                    case SyntaxKind.JSDocPropertyTag:
                        symbol ??= getSymbolOfDeclaration(declaration);
                        const type = getTypeOfSymbol(symbol);
                        return !!(symbol.flags & SymbolFlags.Property && symbol.flags & SymbolFlags.Optional && isOptionalDeclaration(declaration) && (symbol as MappedSymbol).links?.mappedType && containsNonMissingUndefinedType(type));
```

And then that led to obtaining an optional type for `undefined` (that's the `type` in the code above) but we never should try to do that with `strictNullChecks: false`. And in a sense, this proposed fix is solely based on that fact - we don't ever have to "add" `undefined` to anything when `strictNullChecks: false`.

Reply:
I could also fix this by avoiding adding `undefined` to... `undefined` but this seemed cleaner and it potentially skips some extra redundant work (although I perhaps could try to detect this in `requiresAddingImplicitUndefined` too)

Reply:
I don't have a problem with ID depending on compiler options. The compiler options should be known when you transpile the file.

Reply:
Actually, now that I read better, is there a reason why `canReuseTypeNodeAnnotation` isn't just calling `addOptionality`?

The crashing code is:

```ts	
    function getOptionalType(type: Type, isProperty = false): Type {
        Debug.assert(strictNullChecks);
```

But `addOptionality` is the thing we typically use to not assert.

Reply:
Yeah, this works:

```diff
diff --git a/src/compiler/checker.ts b/src/compiler/checker.ts
index b439b7fd9c..ba54e79d69 100644
--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -6266,7 +6266,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                     return true;
                 }
                 if (requiresAddingUndefined && annotationType) {
-                    annotationType = getOptionalType(annotationType, !isParameter(node));
+                    annotationType = addOptionality(annotationType, !isParameter(node));
                 }
                 return !!annotationType && typeNodeIsEquivalentToType(node, type, annotationType) && existingTypeNodeIsNotReferenceOrIsReferenceWithCompatibleTypeArgumentCount(existing, type);
             },
diff --git a/src/compiler/expressionToTypeNode.ts b/src/compiler/expressionToTypeNode.ts
index 43abd33c65..ad4f832106 100644
--- a/src/compiler/expressionToTypeNode.ts
+++ b/src/compiler/expressionToTypeNode.ts
@@ -814,7 +814,7 @@ export function createSyntacticTypeNodeBuilder(
     }
     function typeFromProperty(node: PropertyDeclaration | PropertySignature | JSDocPropertyTag, symbol: Symbol, context: SyntacticTypeNodeBuilderContext) {
         const declaredType = getEffectiveTypeAnnotationNode(node);
-        const requiresAddingUndefined = strictNullChecks && resolver.requiresAddingImplicitUndefined(node, symbol, context.enclosingDeclaration);
+        const requiresAddingUndefined = resolver.requiresAddingImplicitUndefined(node, symbol, context.enclosingDeclaration);
         let resultType = failed;
         if (declaredType) {
             resultType = syntacticResult(serializeTypeAnnotationOfDeclaration(declaredType, context, node, symbol, requiresAddingUndefined));
```

Reply:
Thanks for adding the above patch; the line here can now be removed.